### ISSUE TITLE
MPI scaling issues in solvers (triaged, long term, epic)

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#1275]
@trilinos/tpetra 

Issue reported by @spdomin and currently tracked elsewhere.

Stub summary: Some MPI implementations consumes buffer space on a process proportional to the number of processes with which that process exchanges point-to-point messages.  Buffers grow dynamically, but the MPI implementation annoyingly never returns that memory.  This puts a burden on applications to prefer collectives to point-to-point messages, and in general to keep the in- and out-degree of point-to-point messages on a process low.

**Comment by mhoemmen** [comment]
An e-mail exchange and phone conversation this past week suggests that the cause could be a Map with badly scattered and strided GIDs on some processes, that makes Tpetra::Directory do a lot of point-to-point communication to resolve ownership of GIDs.  It looks like collectives (at least `MPI_Alltoall`) don't have this buffer growth issue, but I'm waiting on Anthony to find out if one-sided MPI communication does.

**Comment by spdomin** [comment]
On a sandy bridge maching running the R3 and R4 mesh, here is the spread of global ids and the matrix row size (max over all rows):

The R3 and R4 study (Serrano 9 and 72 node) is as follows:

R3:

Max range is: 17598828 on this rank: 3
Momentum globalyOwnedMatrix/ownedMatrix size: 78/81
Continuity globalyOwnedMatrix/ownedMatrix size: 26/27
MixFraction globalyOwnedMatrix/ownedMatrix size: 26/27

R4:

Max range is: 140466197 on this rank: 262
Momentum globalyOwnedMatrix/ownedMatrix size: 78/81
Continuity globalyOwnedMatrix/ownedMatrix size: 26/27
MixFraction globalyOwnedMatrix/ownedMatrix size: 26/27

Nalu patch included...

[maxDiagTpetra.txt](https://github.com/trilinos/Trilinos/files/965582/maxDiagTpetra.txt)

Looks like range is rather high while the row size is what is expected for CVFEM.

**Comment by maherou** [comment]
Is there any expectation that mesh or matrix connectivity is symmetric?  Symmetry properties could be useful for analysis.

Also, are these the only two rows with wide range of GIDs?

**Comment by mhoemmen** [comment]
I just posted some proposed fixes here:

https://prod.sandia.gov/sierra-trac/wiki/TrinityAcceptance#StatusreportbyMarkHoemmen

One of the proposals could be novel work, so I didn't want to post it in public.  I also need to check to make sure that revealing specific MPI functions that allocate a lot of memory (besides what we already talked about above) doesn't give away NDA information.

**Comment by spdomin** [comment]
@alanw0, could you subscribe to this issue? I think some of the work suggested above may have already been done (the simple things, that is)..

**Comment by alanw0** [comment]
Wow there's a lot of traffic about this issue, I'm not sure I'm fully caught up...
The stk team did quite a bit of work a couple years ago to improve scalability for some nalu cases. At that time the creation of internal mesh edges was showing up in profiles and we knocked it down quite a bit for runs on ~100k procs. In the last day or two the stk team has fielded a couple of questions from Paul Lin who seems to be looking at scaling issues with initial mesh read/populate, and was digging into some associated stk functions asking if they might be doing all-to-all communications with point-to-point operations. The answer was no, however it could turn into that if the mesh had some pathological connections (a node/edge/face that's connected to many many others).

I'm not sure of the origin of some of the meshes used for nalu scaling studies. If they come from mesh-adapt, I would be interested in some particulars. Such as are the parent elements still present when the refined children are written out? If nalu's input mesh contained parent elements that could complicate connectivity and sharing, which would influence communication in stk-mesh.

**Comment by spdomin** [comment]
@alanw0 - yes, lots of history on this topic. Please start with Mark's link above and think about the solver interface through TpetraLinSys... No need to worry about the past STK-based questions unless they come up again. I would like for us two and Mark to support this effort from the ExaWind perspective.

**Comment by alanw0** [comment]
It looks like Mark is suggesting that the organization of the GIDs being given to Trilinos maps could be improved. I know we are keeping a field 'naluGlobalId_' where we organize the ids, which is simple except for cases where there are periodic bcs or other situations that cause multiple nodes to have the same id. I don't know if those cases exist in the scaling problems that are being run. In any case, I would like to better understand Mark's comments about what would make life easier for Map::getRemoteIndexList. @mhoemmen perhaps we can chat about this.

**Comment by mhoemmen** [comment]
tl;dr "Could help, but let's fix what we can in Trilinos and possibly also STK"

@alanw0 and I chatted about this earlier today.  It might help for Nalu to reorganize the GIDs before handing them off to Trilinos' Maps and solvers.  However, the problem may come up in general, regardless of what Nalu does.  We're also not sure if STK might have the same problem, since it may use the same or similar communication operations for comparable functionality.

@alanw0 and I talked about a separate optimization, which is for STK to pass its (already computed) communication pattern to Trilinos for solves.  This would make a lot of sense for performance reasons.  However, STK and Trilinos should do similar things.  If STK is _not_ causing MPI to eat up memory, then Trilinos should do what STK is doing.  Otherwise, both STK and Trilinos need to change to avoid this issue, and we should continue working together to avoid duplicated effort.

**Comment by mhoemmen** [comment]
This is partly related to #1329 (not in Trilinos -- `MPI_Reduce_scatter` use was elsewhere), but only partly.

**Comment by mhoemmen** [comment]
New back-end for MPI 1-sided will come on line in June; we could experiment with that then.
